storage/range.go

```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
-		cmd.done <- err
-		return err
+		reply = cmd.Reply
+		ctx = cmd.ctx
+	} else {
+		// This command originated elsewhere so we must create a new reply buffer.
+		reply = args.CreateReply()
+		ctx = r.context()
+	}
+
+	// Verify the leader lease is held; Note that we don't require the leader
+	// lease when trying to grant the leader lease!
+	held, expired := verifyLeaderLease(r.getLease(), origin, args.Header().Timestamp)
+	if (!held || expired) && args.Method() != proto.InternalLeaderLease {
+		// This error case is somewhat special: Any command that makes it into
+		// Raft should have the leader lease held by the proposer. The only
+		// case that violates this assumption is that of the proposer having a
+		// corrupted internal state and erroneously assuming leadership. Then,
+		// the command should certainly not be executed by the replicas, but
+		// the error must not enter the response cache (which is updated in
+		// applyRaftCommand below) since it is going to be retried with exactly
+		// the same timestamp. On retry, it may be proposed by the real leader
+		// (or a later leader, which thereby assumes leadership over all past
+		// timestamps as well), for which the command should execute. If it did
+		// hit the response cache, the distributed sender would get stuck in an
+		// infinite loop, retrieving a stale NotLeaderError over and over
+		// again.
+		err = r.newNotLeaderError()
+	} else {
+		err = r.maybeSetCorrupt(
+			r.applyRaftCommand(ctx, index, origin, args, reply),
+		)
 	}
-	// This command originated elsewhere so we must create a new reply buffer.
-	err := r.applyRaftCommand(r.context(), index, proto.RaftNodeID(raftCmd.OriginNodeID), args, args.CreateReply())
-	if err != nil && log.V(1) {
+
+	// TODO(tamird,tschottdorf): according to #1400 we intend to set the reply
+	// header's error as late as possible and in a central location. Range
+	// commands still write to the header directly, but once they don't this
+	// could be the authoritative location that sets the reply error for any-
+	// thing that makes it into Raft. Note that we must set this prior to
+	// signaling cmd.done below, or the waiting RPC handler might proceed
+	// before we've updated its reply.
+	reply.Header().SetGoError(err)
+
+	if cmd != nil {
+		cmd.done <- err
+	} else if err != nil && log.V(1) {
 		log.Errorc(r.context(), "error executing raft command %s: %s", args.Method(), err)
 	}
+
 	return err
 }
 
-// applyRaftCommand applies a raft command from the replicated log to
-// the underlying state machine (i.e. the engine).
-// The caller needs to hold the range lock.
-func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID proto.RaftNodeID, args proto.Request, reply proto.Response) error {
+// applyRaftCommand applies a raft command from the replicated log to the
+// underlying state machine (i.e. the engine). The caller needs to hold the
```

Comment:
Does the caller in fact need to hold the range lock? That doesn't appear to be true.